src/containers/tokens/components/bonding-curve-form/index.js

```diff
@@ -0,0 +1,422 @@
+        >
+          BUY
+        </Button>
+
+        <SellPNKToBondingCurveFormHeader />
+        <div className="Tokens-sell-form">
+          <div>
+            <SellPNKToBondingCurveForm
+              enableReinitialize
+              keepDirtyOnReinitialize
+              initialValues={{
+                rate: (
+                  <div>
+                    <small>
+                      Estimated amount of ETH you'll get: {this.estimateETH()}
+                    </small>
+                    <br />
+                    <small>Exchange rate: 1 ETH = {this.sellPrice()} PNK</small>
+                  </div>
+                )
+              }}
+              onSubmit={this.handleSellPNK}
+              disabled={this.isApproveRequired()}
+            />
+          </div>
+          <div className="Tokens-sell-unlock">
+            {approveTransactionProgress ===
+            'done' ? null : !this.isApproveRequired() ? null : approveTransactionProgress ===
+              'pending' ? (
+              <span>Unlocking...</span>
+            ) : approveTransactionProgress === 'confirming' ? (
+              <span>Confirming...</span>
+            ) : (
+              <Button onClick={this.unlock} className="Tokens-unlock-button">
+                UNLOCK
+              </Button>
+            )}
+          </div>
+        </div>
+        <Button
+          onClick={submitSellPNKToBondingCurveForm}
+          size="small"
+          disabled={!this.isSellButtonEnabled()}
+          className="Tokens-form-button"
+        >
+          SELL
+        </Button>
+
+        <div>
+          <small className="Tokens-form-footer">
+            Exchange is powered by{' '}
+            <a
+              href="https://uniswap.exchange"
+              target="_blank"
+              rel="noopener noreferrer"
+            >
+              UniSwap
+            </a>
+            .
+          </small>
+        </div>
+      </div>
+    )
+  } // render()
+}
+
+export default connect(
+  state => ({
+    inputETH: formValueSelector('buyPNKFromBondingCurveForm')(
+      state,
+      'amountOfETH'
+    ),
+    inputPNK: formValueSelector('sellPNKToBondingCurveForm')(
+      state,
+      'amountOfPNK'
+    ),
+    buyPNKFromBondingCurveFormIsInvalid: getBuyPNKFromBondingCurveFormIsInvalid(
+      state
+    ),
+    sellPNKToBondingCurveFormIsInvalid: getSellPNKToBondingCurveFormIsInvalid(
+      state
+    ),
+    approveTransactionProgress:
+      state.bondingCurve.approveTransactionProgress.data
+  }),
+  {
+    getSellPNKToBondingCurveFormIsInvalid,
+    submitSellPNKToBondingCurveForm,
+    getBuyPNKFromBondingCurveFormIsInvalid,
+    submitBuyPNKFromBondingCurveForm,
+    buyPNKFromBondingCurve: bondingCurveActions.buyPNKFromBondingCurve,
+    sellPNKToBondingCurve: bondingCurveActions.sellPNKToBondingCurve,
+    approvePNKToBondingCurve: bondingCurveActions.approvePNKToBondingCurve
+  }
+)(BondingCurveForm)
+
+const SPREAD_FACTOR = toBN(997)
```

Comment:
These functions feel out of place outside of the component. Why not pass spread factor and spread divisor as params and make these `utils`?

Reply:
These functions are only used by this component and there is no foreseeable chance to use them anywhere. I think `utils` should contain common routines used at multiple places.

Reply:
ok